node/libs/crypto/src/secp256k1/mod.rs

```diff
@@ -0,0 +1,278 @@
+        let (r, s) = self.sig.split_bytes();
+        bz[..32].copy_from_slice(&r);
+        bz[32..64].copy_from_slice(&s);
+        bz[64] = self.recid.to_byte() + 27;
+        bz
+    }
+}
+
+impl Hash for Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write(&self.encode())
+    }
+}
+
+impl PartialOrd for Signature {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Signature {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        ByteFmt::encode(self).cmp(&ByteFmt::encode(other))
+    }
+}
+
+/// Seclp256k1 multi-sig
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+pub struct AggregateSignature(Vec<Signature>);
+
+impl AggregateSignature {
+    /// Add a signature to the aggregate.
+    pub fn add(&mut self, sig: Signature) {
+        self.0.push(sig)
+    }
+
+    /// Verifies an aggregated signature for multiple messages against the provided list of public keys.
+    ///
+    /// Verification fails if the total number of messages and signatures do not match.
+    ///
+    /// This method was added to mimic the bn254 version, which used BLS signature aggregation and did
+    /// not rely on message and key ordering; it supported different messages from each signatory as well.
+    ///
+    /// Here we have a simple list of signatures, and we cannot rely on them having been added in the same
+    /// order in which they are going to be verified, therefore we have to try every message against every
+    /// signature.
+    ///
+    /// The method assumes that there are no repeated pairs in the input, ie. that every signature is used exactly once.
+    pub fn verify_hash<'a>(
+        &self,
+        hashes_and_pks: impl Iterator<Item = (&'a [u8], &'a PublicKey)>,
+    ) -> anyhow::Result<()> {
+        // Keep track of which signatures have been verified.
+        let mut verified = HashSet::new();
+
+        'inputs: for (i, (hash, pk)) in hashes_and_pks.enumerate() {
+            'signatures: for (j, sig) in self.0.iter().enumerate() {
+                if verified.contains(&j) {
+                    continue 'signatures;
+                }
+                if sig.verify_hash(hash, pk).is_ok() {
+                    verified.insert(j);
+                    continue 'inputs;
+                }
+            }
+            bail!("failed to verify message {i} against any of the signatures in the aggregate");
+        }
+        if verified.len() < self.0.len() {
+            bail!(
+                "not enough messages to verify aggregated signature: expected {}, got {}",
+                self.0.len(),
+                verified.len()
+            );
+        }
+        Ok(())
+    }
+
+    /// Verify messages after hashing them with Keccak256
+    pub fn verify<'a>(
+        &self,
+        msgs_and_pks: impl Iterator<Item = (&'a [u8], &'a PublicKey)>,
```

Comment:
ditto

Reply:
NB this is how it is in [bn254](https://github.com/matter-labs/era-consensus/blob/90b893ceead1aabff34713a152376fffbf2b9e0a/node/libs/crypto/src/bn254/mod.rs#L281-L283)

Reply:
oh right. As long as we make the generic call the non-generic version, we are good.